data/src/main/java/co/nimblehq/compose/crypto/data/ResponseMapping.kt

```diff
@@ -0,0 +1,41 @@
+package co.nimblehq.compose.crypto.data
+
+import co.nimblehq.compose.crypto.data.model.response.ErrorResponse
+import co.nimblehq.compose.crypto.data.service.JsonApiException
+import co.nimblehq.compose.crypto.data.service.UnknownException
+import co.nimblehq.compose.crypto.data.service.providers.MoshiBuilderProvider
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.transform
+import retrofit2.Response
+
+@OptIn(FlowPreview::class)
+fun <T> Flow<Response<T>>.transform(): Flow<T> = transform { response ->
+    flow {
+        val body = response.body()
+        if (response.isSuccessful && body != null) {
+            emit(body)
+        } else {
+            error(mapError(response))
+        }
+    }
+}
+
+private fun <T> mapError(response: Response<T>?): Exception {
+    val errorBody = response?.errorBody()?.string()
+    val errorResponse = parseErrorResponse(errorBody)
+
+    return if (errorResponse != null && response != null) JsonApiException(errorResponse, response)
+    else UnknownException
+}
+
+private fun parseErrorResponse(source: String?): ErrorResponse? {
+    return try {
+        val moshi = MoshiBuilderProvider.moshiBuilder.build()
```

Comment:
Hmm, We already provide `MoshiBuilderProvider.moshiBuilder.build()` in `MoshiModule` Is it duplicate? 🤔

Reply:
@Wadeewee We can't inject the `moshiBuilder` instance into the extension so that's why I have to re-define it here.

Reply:
@hoangnguyen92dn As we don't handle any errors as you said in this [comment](https://github.com/nimblehq/jetpack-compose-crypto/pull/23#discussion_r946620768), but you prefer to leave it. That's why we need to redefine it here. So why don't we remove it if It's unnecessary for now? I think we can implement it later if needed in future

Reply:
@Wadeewee I mean that we don't use the `ErrorResponse` to mapping API response for now but we could leave it as the base flow implementation. We could update the `ErrorResponse` to match the response from API without adding any further stuff.

Reply:
OK, I got it @hoangnguyen92dn. Let's see when we work on the integration task if we needed it 🙏